cpp/test/IceStorm/createOrRetrieve/Client.cpp

```diff
@@ -0,0 +1,192 @@
+        catch (const NoSuchTopic&)
+        {
+            // Expected - topic doesn't exist
+        }
+
+        // Create topic using createOrRetrieve
+        auto topic1 = manager->createOrRetrieve(topicName);
+        test(topic1);
+        test(topic1->getName() == topicName);
+
+        // Verify topic was created
+        auto verifyTopic = manager->retrieve(topicName);
+        test(verifyTopic);
+        test(verifyTopic->getName() == topicName);
+
+        // Clean up
+        topic1->destroy();
+    }
+
+    //
+    // Test 2: Retrieving an existing topic via createOrRetrieve
+    //
+    {
+        const string topicName = "createOrRetrieve-topic2";
+
+        // Create topic first using regular create
+        auto originalTopic = manager->create(topicName);
+        test(originalTopic);
+
+        // Now use createOrRetrieve - should retrieve the existing topic
+        auto retrievedTopic = manager->createOrRetrieve(topicName);
+        test(retrievedTopic);
+        test(retrievedTopic->getName() == topicName);
+
+        // Verify both proxies refer to the same topic
+        test(originalTopic->getName() == retrievedTopic->getName());
+
+        // Clean up
+        originalTopic->destroy();
+    }
+
+    //
+    // Test 3: Concurrent calls to createOrRetrieve for the same topic name
+    //
+    {
+        const string topicName = "createOrRetrieve-topic3";
+
+        // Ensure topic doesn't exist
+        try
+        {
+            auto existingTopic = manager->retrieve(topicName);
+            existingTopic->destroy();
+        }
+        catch (const NoSuchTopic&)
+        {
+            // Expected - topic doesn't exist
+        }
+
+        // Launch multiple concurrent createOrRetrieve calls
+        const int numThreads = 10;
+        vector<future<optional<TopicPrx>>> futures;
+
+        for (int i = 0; i < numThreads; ++i)
+        {
+            futures.push_back(async(
+                launch::async,
+                [&manager, &topicName]()
+                {
+                    try
+                    {
+                        return manager->createOrRetrieve(topicName);
+                    }
+                    catch (const exception& ex)
+                    {
+                        cerr << "Exception in concurrent createOrRetrieve: " << ex.what() << endl;
+                        throw;
+                    }
+                }));
+        }
+
+        // Collect all results
+        vector<optional<TopicPrx>> topics;
+        for (auto& f : futures)
+        {
+            auto topic = f.get();
+            test(topic);
+            test(topic->getName() == topicName);
+            topics.push_back(topic);
+        }
+
+        // All topics should have the same name
+        for (const auto& topic : topics)
+        {
+            test(topic->getName() == topicName);
+        }
+
+        // Verify only one topic was actually created
+        auto finalTopic = manager->retrieve(topicName);
```

Comment:
I would call retrieveAll and check there is a single topic with the correct name.